idea/src/org/jetbrains/kotlin/idea/inspections/RedundantElseInIfInspection.kt

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2010-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.idea.inspections
+
+import com.intellij.codeInspection.LocalQuickFix
+import com.intellij.codeInspection.ProblemDescriptor
+import com.intellij.codeInspection.ProblemHighlightType
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiDocumentManager
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiElementVisitor
+import com.intellij.psi.codeStyle.CodeStyleManager
+import org.jetbrains.kotlin.idea.caches.resolve.analyze
+import org.jetbrains.kotlin.idea.intentions.branchedTransformations.isElseIf
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.endOffset
+import org.jetbrains.kotlin.psi.psiUtil.getStrictParentOfType
+import org.jetbrains.kotlin.psi.psiUtil.startOffset
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.types.typeUtil.isNothing
+
+class RedundantElseInIfInspection : AbstractKotlinInspection() {
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor =
+        ifExpressionVisitor(fun(ifExpression) {
+            if (ifExpression.isElseIf()) return
+            val elseKeyword = ifExpression.singleElseKeyword(true) ?: return
+            val rangeInElement = elseKeyword.textRange?.shiftRight(-ifExpression.startOffset) ?: return
+            holder.registerProblem(
+                holder.manager.createProblemDescriptor(
+                    ifExpression,
+                    rangeInElement,
+                    "Redundant 'else'",
+                    ProblemHighlightType.LIKE_UNUSED_SYMBOL,
+                    isOnTheFly,
+                    RemoveRedundantElseFix()
+                )
+            )
+        })
+}
+
+private class RemoveRedundantElseFix : LocalQuickFix {
+    override fun getName() = "Remove redundant 'else'"
+
+    override fun getFamilyName() = name
+
+    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
+        val ifExpression = descriptor.psiElement as? KtIfExpression ?: return
+        val elseKeyword = ifExpression.singleElseKeyword(false) ?: return
+        val elseExpression = elseKeyword.getStrictParentOfType<KtIfExpression>()?.`else` ?: return
+
+        val copy = elseExpression.copy()
+        if (copy is KtBlockExpression) {
+            copy.lBrace?.delete()
+            copy.rBrace?.delete()
+        }
+        val added = ifExpression.parent.addAfter(copy, ifExpression)
+        elseExpression.delete()
+        elseKeyword.delete()
+        val nextSibling = added.nextSibling
+
+        if (nextSibling != null) {
```

Comment:
I don't like this piece of code. Are you sure it's required here? Please explain why.

Reply:
I would like to indent after deleting `else` block. Without this code, it looks like this:

before
```
fun test(x: Boolean) {
    if (x) {
        return
    } else<caret> {
        foo()
        bar()
    }
}
```

after
```
fun test(x: Boolean) {
    if (x) {
        return
    }
    foo()
        bar()
}
```

Reply:
Looks like it does not work as you want, see my title comment.